clients/src/test/java/org/apache/kafka/clients/consumer/internals/FetcherTest.java

```diff
@@ -2421,6 +2413,93 @@ public void testGetOffsetsFencedLeaderEpoch() {
         assertEquals(0L, metadata.timeToNextUpdate(time.milliseconds()));
     }
 
+    @Test
+    public void testGetOffsetByTimeWithPartitionsRetryCouldTriggerMetadataUpdate() {
+        List<Errors> retriableErrors = Arrays.asList(Errors.NOT_LEADER_FOR_PARTITION,
+            Errors.REPLICA_NOT_AVAILABLE, Errors.KAFKA_STORAGE_ERROR, Errors.OFFSET_NOT_AVAILABLE,
+            Errors.LEADER_NOT_AVAILABLE, Errors.FENCED_LEADER_EPOCH, Errors.UNKNOWN_LEADER_EPOCH);
+
+        final int newLeaderEpoch = 3;
+        MetadataResponse updatedMetadata = TestUtils.metadataUpdateWith("dummy", 3,
+            singletonMap(topicName, Errors.NONE), singletonMap(topicName, 4), tp -> newLeaderEpoch);
+        LogContext dummyContext = new LogContext();
+        ConsumerMetadata dummyMetadata = new ConsumerMetadata(0, Long.MAX_VALUE, false, false,
```

Comment:
`MetadataResponse` allows us to get the `Cluster` directly, so we can do something simpler:
```
        Node oldLeader = initialUpdateResponse.cluster().leaderFor(tp1);
        Node newLeader = updatedMetadata.cluster().leaderFor(tp1);
        assertNotEquals(oldLeader, newLeader);
```
Since the metadata doesn't change, we can just do this check once.

Reply:
Sounds good